### DISCUSSION TITLE
tools used ?

### DISCUSSION THREAD

**Discussion by mirsella** [timschneeb/GalaxyBudsClient#290] [General]
hey

first thanks for your work !
I have Nothing ear (1) buds and would like to make my own app to control the earbuds audio cancellation mode.

to reverse engineering the APK I guess you used apktool / jadx, but It seems like you also just used something to intercept the Bluetooth trafic.
the only project I could find was btproxy which didn't seemed to work for me.

can you share the tools you have used and maybe a little insight on how ? thanks

**Comment by timschneeb** [comment]
Hey,
Yes, for decompilation, I'm using jadx.

For Bluetooth traffic interception, however, I wrote my custom C# program because I needed proper extensibility to add parsers for the encoded binary messages, and I also needed to send a bunch of custom binary data for testing. So I can't really point you to a specific tool. At the time, I used https://github.com/inthehand/32feet for interacting with Bluetooth devices and opening a serial socket on them. Then I started using that C# program as a foundation for GalaxyBudsClient itself.

Also, instead of collecting on your PC using btproxy or similar tools, you should do it right on your Android phone (or wherever the original app is installed).
To intercept Bluetooth traffic using your Android phone, consider using the HCI dump feature (some manufacturers block this option; you might need root in that case): https://stackoverflow.com/a/30352487

**Reply by mirsella** [reply]
thanks a lot, forgot about hci logs directly from android because the file wasn't created in /sdcard for me, and exists in /data/misc/bluetooth/logs only when bluetooth is activated, after the file is deleted. i'm searching now with wireshark.

**Comment by mirsella** [comment]
update : i'm abandoning lol. managed to find some things data which could be what i'm searching for (ANC on/off, transparency mode) but i can't manage to find a way to replicate the data sent to my earbuds from my pc
